table/table.cpp

```diff
@@ -164,59 +133,120 @@ void Table::edit_rows(vector<string> &columns,
             Info index = column_index(column);
             edit_indecies.push_back(index);
         } catch (std::exception e) {
-            // catch code
-            // throw()
+            throw e;
         }
     }
+
+    // Loops through rows and edits the ones that fit comparisons
     for(size_t i = 0; i < table.rows_size(); ++i) {
-        bool good = true;
-        for(size_t j = 0; j < comparisons.size(); ++j) {
-            bool correct_entry = correct_type(indecies[j], table.rows(i).entries(indecies[j].index));
-            bool correct_comparison = correct_type(indecies[j], comparisons[j]);
-            if(!correct_entry || !correct_comparison) {
-                // throw()
-            }
-            if(!compare_entries(table.rows(i).entries(indecies[j].index), comparisons[j])) {
-                good = false;
-                j = comparisons.size();
-            }
-        }
+        bool good = check_row(table.rows(i), indecies, comparisons);
+
         if(!good) {
             continue;
         }
+
+        // Edits each specified entry in row
         for(size_t j = 0; j < entries.size(); ++j) {
             if(!correct_type(edit_indecies[j], entries[j])) {
-                // throw
+                throw type_mismatch(type_mismatch::error_type::updation, get_type(edit_indecies[j]), get_type(entries[j]));
             }
             ::database::Entry* edit_entry = table.mutable_rows(i)->mutable_entries(edit_indecies[j].index);
             edit_entry->CopyFrom(entries[j]);
         }
     }
 } // edit_rows()
 
-void Table::insert(::database::Row &row) {
-    for(auto it : column_indecies) {
+// Edits all rows to set entry@each edit_column = the said new entry
+// Input: vector<string> &edit_columns -> columns being edited in filtered entries
+// Input: vector<::database::Entry> &entries -> new entries to update entries in edit_columns
+void Table::edit_all(vector<string> &edit_columns, 
+                vector<::database::Entry> &entries) {
+    vector<Info> edit_indecies;
+    for (string column: edit_columns) {
         try {
-            if(!correct_type(it.second, row.entries(it.second.index))) {
-                throw out_of_range("Row is wrong type");
+            Info index = column_index(column);
+            edit_indecies.push_back(index);
+        } catch (std::exception e) {
+            throw e;
+        }
+    }
+
+    for(size_t i = 0; i < table.rows_size(); ++i) {
+        // Edits each specified entry in row
+        for(size_t j = 0; j < entries.size(); ++j) {
+            if(!correct_type(edit_indecies[j], entries[j])) {
+                throw type_mismatch(type_mismatch::error_type::updation, get_type(edit_indecies[j]), get_type(entries[j]));
```

Comment:
Suppose we get to this case where there is a type_mismatch in the middle of the for loop (rather than at first).

Then we would potentially edit some entries that were successful before the exception was thrown, but would not complete editing entries after the exception, correct?

This would leave the table in an undefined state (some entries are edited, some aren't) to the user. IMO, I think the correct behavior should be that all type checks/validations are complete first and then editing entries are done after all validation is passed so that the user/client/caller of this code knows what is the state of the table before making subsequent operations on it.